keps/sig-network/2030-endpointslice-subsetting/README.md

```diff
@@ -0,0 +1,264 @@
+```
+
+In the future this pattern may be expanded to include other concepts or
+topologies. A simple pattern like this will allow EndpointSlices to be delivered
+to consumers in a specific zone or region.
+
+
+### Risks and Mitigations
+
+This approach does not allow a single EndpointSlice to target multiple zones or
+regions. Any approach that enabled that would be significantly more complicated.
+The initial proposal in [KEP
+#2004](https://github.com/kubernetes/enhancements/issues/2004) suggests that
+this won't be an issue. 
+
+## Design Details
+
+### API Consumption
+
+EndpointSlice API consumers can use the following selectors to consume
+EndpointSlices:
+
+#### Select EndpointSlices Without Labels
+This is required to support producers of EndpointSlices that don't set these
+labels, including older versions of the EndpointSlice controller.
+
+```
+matchExpressions:
+  - {key: endpointslice.kubernetes.io/for-zone, operator: DoesNotExist}
+  - {key: endpointslice.kubernetes.io/for-region, operator: DoesNotExist}
+```
+
+#### Select EndpointSlices With Matching Zone
+```
+matchLabels:
+  endpointslice.kubernetes.io/for-zone: example-zone
+```
+
+#### Select EndpointSlices With Matching Region
+```
+matchLabels:
+  endpointslice.kubernetes.io/for-region: example-region
+```
+
+#### Kube-Proxy
+When the `EndpointSliceSubsetting` feature gate is set to true on `Kube-Proxy`,
+it will use these selectors to filter EndpointSlices.
+
+### Controller Implementation
+Although a controller implementation is out of scope for this KEP, it is worth
+discussing what that might look like. For reference, [KEP
+#2004](https://github.com/kubernetes/enhancements/issues/2004) discusses how this
+could be implemented for the EndpointSlice controller. That proposal involves 3
+potential approaches - Original, PreferZone, and RequireZone.
+
+None of the proposed approaches would involve data duplication. Each
+Pod/endpoint would continue to live in a single EndpointSlice. The reason they
+might end up with more EndpointSlices would be less efficient packing. Here's
+the number of EndpointSlices that would result based on the number of endpoints
+a Service has in a 3 zone cluster.
+
+| # endpoints | Original # slices | PreferZone # slices | RequireZone # slices |
+|-|-|-|-|
+| 6 | 1 | 1 | 3 |
+| 80 | 1 | 3 | 3 |
+| 250 | 3 | 3 | 3 |
+
+The RequireZone approach requires at least one EndpointSlice per zone per
+Service. The PreferZone also has the same requirement unless the minimum
+threshold has been met. Before that threshold, a single shared EndpointSlice (no
+additional labels) is used. There's some padding involved here to make sure
+we're not flapping back and forth between these states.
+
+With this approach, EndpointSlices can be delivered everywhere (no additional
+labels), or to a zone (for-zone), or to a region (for-region). None of the
+proposed approaches involve a single Service having separate sets of
+EndpointSlices for each use case.
```

Comment:
Should this xref https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/1659-standard-topology-labels specifically: `"region" and "zone" are strictly hierarchical ("zones" are subsets of "regions") and that zone names are unique across regions`

Reply:
Good call, added a reference to this.